src/read/mod.rs

```diff
@@ -339,9 +449,11 @@ impl<R: AsyncRead + Unpin> AsyncRead for CompressionReader<R> {
 }
 
 impl<'a, R: AsyncRead + Unpin> CompressionReader<R> {
-    pub(crate) fn from_reader(compression: &Compression, reader: Take<R>) -> Self {
-        match compression {
-            Compression::Stored => CompressionReader::Stored(reader),
+    pub(crate) fn from_reader(compression: &Compression, reader: R, take: Option<u64>) -> Result<Self> {
+        Ok(match compression {
+            Compression::Stored => CompressionReader::Stored(
+                BufReader::new(reader).take(take.ok_or_else(|| ZipError::MissingCompressedSize)?),
```

Comment:
I don't think this error can ever be triggered since, from what I can tell, we're always passing in a Some.
There probably needs to be some logic somewhere (probably stream read_lfh) that detects if data descriptor bit is set and compressed_size is 0 and map crc and sizes to None.

Reply:
You're right, I totally forgot this (I've been through this logic a bunch of times). But I'm unsure how to properly do this. Empty files would have `crc32 = 0`, as well as `compressed_size = 0` and `uncompressed_size = 0`.

> You can try to create an empty file: `touch foo` and zip it: `7z a foo.zip foo` (or `zip foo.zip foo`), and those values will all be zero.

I can only think of setting both to None if they are all `0` and `header.flags.data_descriptor = true`, so this would correctly yield the error. 

The only problem is, **rs-async-zip** sets `header.flags.data_descriptor` to `true` when using `write_entry_stream`, even for `Stored` mode. But `ZipFileWriter` uses a single-pass zip creation, which means that it does not *Seek* back to write the size once it's done writing the data. This is not a problem when using a *Reader* that does support *Seek* because we just *Seek* to the *Central Directory* and take the metadata from there.

I can think of two workarounds:

- A *Seek* *ZipFileWriter* which can go back and write the sizes and crc
  - But it would need to be an specialized writer, so we can still do a single-pass zip creation.
- Add an `write_sized_entry_stream` (I don't like this name) which could be used to write N bytes from the stream, using `Take<T>`
  - This guarantees that we only write N bytes, and we can write the number of bytes beforehand, we don't need the CRC32 to know how much data we need to read, so we can write 0 here (unsure how others decoders would handle this, which is bad, would need some testing)
  - Combined with the previous approach, we can go back and write the CRC32 to conform to the spec
  - This is useful for things like `tokio::fs::File`, since if we successfully retrieve the `Metadata`, we can get the `size` without even reading the file. The only problem is that, files are mutable, if the size increases, that is not a problem, we will be using `Take<T>`, but if it shrinks, we would need to error out.

Reply:
> I can only think of setting both to None if they are all 0 and header.flags.data_descriptor = true, so this would correctly yield the error.

👍 

> A Seek ZipFileWriter which can go back and write the sizes and crc

I want to provide as much functionality so having a one-pass and a two-pass seekable version of the writer seems like a good idea - probably best to be a separate PR though (and potentially further down the line).